File: packages/web/src/lib/taskClassificationDefaults.ts
import type { BoardCadence, DeliveryMode, GovernanceClass, Methodology } from '@/types';

/**
 * Methodology-aware defaults for a new task's governance/delivery taxonomy (#2667).
 *
 * `governance_class` and `delivery_mode` used to always default to `'flow'` /
 * `'waterfall'` regardless of the project's own resolved methodology — a
 * self-contradictory dialog on a WATERFALL project (Flow governance describes
 * itself as "agile, sprint- or kanban-governed work" on a project with no
 * sprints or board governance) and a correctness bug on AGILE (delivery mode
 * stuck at waterfall never engages point-burndown rollup or agile-aware Monte
 * Carlo).
 *
 * The server is the source of truth (API-first): the task-create serializer
 * resolves the same default when a client omits the field, so mobile and MCP
 * agree with web. This module mirrors that resolution purely so the dialog can
 * open on the value the server would pick — there is no "preview the default"
 * endpoint to read it back from. Keep in sync with
 * `packages/api/src/trueppm_api/apps/projects/task_classification_defaults.py`.
 */

/**
 * The governance class a new task should default to for a project's preset.
 *
 * `WATERFALL` → `gated` (phase-gate governance matches a project with no
 * sprints or board governance). Every other preset (`AGILE`, `HYBRID`) keeps
 * the existing `flow` default — `HYBRID` deliberately, because it is the
 * preset that mixes both governance overlays and has no single "right" class
 * to steer toward.
 */
export function defaultGovernanceClassFor(methodology: Methodology): GovernanceClass {
  return methodology === 'WATERFALL' ? 'gated' : 'flow';
}

/**
 * The delivery mode a new task should default to for a project's preset.
 *
 * - `WATERFALL` / `HYBRID` → `waterfall` (`HYBRID` keeps today's literal
 *   default — the lossless choice for a preset that mixes delivery modes).
 * - `AGILE` → `scrum`, unless the project's board already runs continuous-flow
 *   Kanban with no sprint cadence (`boardCadence === 'continuous'`, ADR-0164),
 *   in which case → `kanban` — the deliberate flow-without-sprints choice.
 *   `scrum` is the load-bearing default for a brand-new agile project with
 *   nothing configured yet: an unpointed scrum task degrades to identical
 *   rollup behavior as a waterfall one, while kanban would immediately
 *   binarize percent-complete and equal-weight every task (see #2667).
 */
export function defaultDeliveryModeFor(
  methodology: Methodology,
  boardCadence: BoardCadence,
): DeliveryMode {
  if (methodology === 'AGILE') {
    return boardCadence === 'continuous' ? 'kanban' : 'scrum';
  }
  return 'waterfall';
}

/** A methodology-consistent subset of a taxonomy field's values, and the rest. */
export interface TaxonomyGroups<V extends string> {
  /** Values consistent with the project's own preset — shown first. */
  primary: V[];
  /** Every other value — still selectable, just de-emphasized (never removed;
   *  the taxonomy is a documented hybrid seam, ADR-0036). */
  other: V[];
}

/**
 * How to group `governance_class` options for a project's preset, or `null`
 * to render the flat, ungrouped list.
 *
 * `HYBRID` returns `null` — it is the preset that deliberately mixes both
 * governance overlays, so there is no "off-preset" value to de-emphasize.
 */
export function governanceClassGroups(
  methodology: Methodology,
): TaxonomyGroups<GovernanceClass> | null {
  switch (methodology) {
    case 'WATERFALL':
      return { primary: ['gated'], other: ['flow', 'hybrid'] };
    case 'AGILE':
      return { primary: ['flow'], other: ['gated', 'hybrid'] };
    default:
      return null;
  }
}

/**
 * How to group `delivery_mode` options for a project's preset, or `null` to
 * render the flat, ungrouped list.
 *
 * Both `scrum` and `kanban` count as on-preset for `AGILE` — the split between
 * them is a board-cadence choice, not a methodology mismatch. `milestone`
 * always sits in `other`: it is driven by the `is_milestone` toggle rather
 * than chosen directly, and only ever appears on this control when editing an
 * existing milestone task.
 */
export function deliveryModeGroups(
  methodology: Methodology,
): TaxonomyGroups<DeliveryMode> | null {
  switch (methodology) {
    case 'WATERFALL':
      return { primary: ['waterfall'], other: ['scrum', 'kanban', 'milestone'] };
    case 'AGILE':
      return { primary: ['scrum', 'kanban'], other: ['waterfall', 'milestone'] };
    default:
      return null;
  }
}

File: packages/api/src/trueppm_api/apps/projects/task_classification_defaults.py
"""Methodology-aware defaults for a new task's governance/delivery taxonomy (#2667).

``Task.governance_class`` and ``Task.delivery_mode`` (ADR-0036, #407) are purely
additive fields with model-level defaults of ``FLOW`` / ``WATERFALL`` — chosen so
every pre-existing row's semantics stayed unchanged when the taxonomy shipped.
Those literals are also what a brand-new task got on *every* project regardless of
its own resolved methodology, which produces a self-contradictory dialog on a
``WATERFALL`` project (opens on Flow governance — "Agile, sprint- or
kanban-governed work" — for a project with no sprints or board governance) and a
correctness bug on ``AGILE`` (delivery mode stays Waterfall, so the task never
engages point-burndown rollup or agile-aware Monte Carlo).

This module resolves the *project-aware* default, consulted only when the client
omits the field on create (:meth:`TaskSerializer.validate`) — an explicit value
always wins, on create and update alike, so the documented hybrid seam (a single
project deliberately holding tasks in different delivery modes/governance
classes) is untouched. ``HYBRID`` keeps today's literal default (``flow`` /
``waterfall``) since it is the preset that already mixes both models.

Mirrored client-side in ``packages/web/src/lib/taskClassificationDefaults.ts`` so
the create dialog opens on the value this module would pick — the two are
independently maintained (there is no "preview the server default" endpoint) and
must be changed together. Keep them in sync.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from trueppm_api.apps.projects.models import Project


def resolve_default_governance_class(effective_methodology: str) -> str:
    """The governance class a new task should default to for a project's preset.

    ``WATERFALL`` → ``gated`` (phase-gate governance matches a project with no
    sprints or board governance). Every other preset (``AGILE``, ``HYBRID``)
    keeps the existing ``flow`` default — ``HYBRID`` deliberately, because it is
    the preset that mixes both governance overlays and has no single "right"
    class to steer toward.
    """
    from trueppm_api.apps.projects.models import GovernanceClass, Methodology

    if effective_methodology == Methodology.WATERFALL:
        return str(GovernanceClass.GATED)
    return str(GovernanceClass.FLOW)


def resolve_default_delivery_mode(effective_methodology: str, board_cadence: str) -> str:
    """The delivery mode a new task should default to for a project's preset.

    - ``WATERFALL`` / ``HYBRID`` → ``waterfall`` (HYBRID keeps today's literal
      default — the lossless choice for a preset that mixes delivery modes).
    - ``AGILE`` → ``scrum``, unless the project's board runs continuous-flow
      Kanban with no sprint cadence (``board_cadence == 'continuous'``), in which
      case → ``kanban``.

    The scrum/kanban split is resolved from ``board_cadence`` rather than an
    extra board-config lookup: ADR-0164 already models "sprint cadence vs.
    continuous flow" as that exact field, and ``scrum`` is the load-bearing
    choice for a brand-new agile project with nothing configured yet — an
    unpointed ``scrum`` task degrades to identical rollup behavior as a
    ``waterfall`` one (``COALESCE(story_points, duration)``), while defaulting to
    ``kanban`` would immediately binarize percent-complete and equal-weight every
    task regardless of size (see #2667 for the full analysis).
    """
    from trueppm_api.apps.projects.models import BoardCadence, DeliveryMode, Methodology

    if effective_methodology == Methodology.AGILE:
        if board_cadence == BoardCadence.CONTINUOUS:
            return str(DeliveryMode.KANBAN)
        return str(DeliveryMode.SCRUM)
    return str(DeliveryMode.WATERFALL)


def resolve_default_classification(
    project: Project, *, workspace: object | None = None
) -> tuple[str, str]:
    """``(governance_class, delivery_mode)`` defaults for a new task on ``project``.

    Resolves ``project``'s ``effective_methodology`` (workspace → program →
    project precedence, ADR-0107) once and derives both fields from it, so a
    caller needing both never resolves methodology twice. ``workspace`` is
    accepted so a caller iterating many tasks against the same project can pass
    the already-loaded singleton (mirrors the other resolvers in this app);
    otherwise it is loaded here.
    """
    from trueppm_api.apps.projects.methodology import resolve_effective_methodology

    effective_methodology = resolve_effective_methodology(project, workspace=workspace)  # type: ignore[arg-type]
    return (
        resolve_default_governance_class(effective_methodology),
        resolve_default_delivery_mode(effective_methodology, project.board_cadence),
    )
